reference-delegate-explainer.md

```diff
@@ -0,0 +1,555 @@
+
+* [Form-Associated Custom Elements](https://html.spec.whatwg.org/dev/custom-elements.html#form-associated-custom-element) can participate in forms like a builtin input.
+* [delegatesFocus](https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus) allows a component to work better with keyboard navigation.
+
+However, there are still missing pieces that prevent a web component from truly being a drop-in replacement for a built-in, including:
+
+* Can't create ID reference links to elements inside the shadow DOM.
+* Can't use built-in attributes like `aria-label` or `role` on the host and have them apply to an element inside the shadow DOM.
+* Non-trivial amount of code required to hook up custom attributes on the host to ARIAMixin attributes on an element inside the shadow DOM.
+* Can't get form-association for "free" by delegating to an input inside.
+
+This proposal solves only the first problem of ID references, and leaves the other problems to be solved by other features. While all of the problems may seem related, they can be designed separately.
+
+## Proposal
+
+### Reference Delegate
+
+Reference Delegate is a new feature that enables creating ARIA links to elements inside a component's shadow DOM, while maintaining encapsulation of the internal details of the shadow DOM.
+
+**Goals**
+
+- Solve only the "missing piece" of cross-root ARIA: how to handle IDREF attributes referring into the shadow DOM. Avoid scope creep.
+- Create a mechanism for ID reference attributes like `aria-activedescendant` and `for` to refer to an element inside a component's shadow DOM.
+- Should be compatible with the ARIAMixin attributes such as `ariaActiveDescendantElement`, to create a complete solution to cross-root ARIA.
+- Should work the same for both closed and open shadow roots.
+- Shadow DOM encapsulation should be preserved: No direct access to any elements inside the shadow tree, and no implementation details leaked into a web component's API.
+- Should allow creating references into multiple nested shadow roots, and across "sibling" shadow roots that don't have a direct parent/child relationship.
+- The solution should be serializable, i.e. support declarative syntax that is expressible in HTML only.
+
+**Non-Goals**
+
+- This is scoped to only solve the problem of referring _into_ the shadow DOM. It relies on ARIAMixin to refer _out_ of the shadow DOM.
+- This feature does not solve the [bottleneck effect](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#limitations-of-these-apis). It is difficult to find a compelling real-world example where this is a problem.
+- This does not affect the how attributes set on the host element work. For example, this does not tackle the problem of forwarding `role` or `aria-label`, etc. from the host element to an element inside.
+
+### The `referenceDelegate` attribute
```

Comment:
Do you think naming these attributes `delegatesReference` and `delegatesReferenceMap` makes sense alongside `delegatesFocus`?

Reply:
I like the name `delegatesReference`, and that it is parallel to `delegatesFocus`. However, it does _sound_ more like it should be a boolean true/false, rather than the ID of the target. (E.g. it should be `delegatesReferenceTo` if it's the ID of the target, but I don't like that option). The name `delegatesReferenceMap` is also a little awkward IMO.

I took another look at the names used for [IDREF attributes](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#supported-attributes) to see if there was possible inspiration there... The naming of the ARIA attributes isn't very consistent: some are verbs (`aria-controls`), some are nouns (`aria-activedescendant`), and some are prepositional phrases (`aria-labelledby`). However, several of the newer attributes follow the pattern of ending with "target": `popovertarget`, `invoketarget`, `interesttarget`.

So, another possible option is `referenceTarget` / `referenceTargetMap`. WDYT?

Reply:
I like that since it helps enforce a common pattern using the "target" suffix.

Another option could be `idReferenceTarget`/`idReferenceTargetMap` or `idRefTarget`/`idRefTargetMap` to emphasize the relationship to IDREF attributes.

Reply:
I've updated the proposal to the `referenceTarget`/`referenceTargetMap` name.